### ISSUE TITLE
Build-Parallel affected by a PowerShell concurrency bug

### ISSUE THREAD

**Issue by codepic** [nightroman/Invoke-Build#180]
A PowerShell bug has been open since 2017 and it affects `Invoke-Build`'s `Build-Parallel`.

Issue: [Engine event processing bypasses "ShouldQueueAndProcessInExecutionThread" causing state corruption and crash](https://github.com/PowerShell/PowerShell/issues/4003)

This will result in 2 threads running in the same Runspace and changing its state simultaneously causing a deadlock or state corruption and crash. A telltale sign of this issue is receiving a `Stack empty.` error but due to the nature of the bug, the results may be anything from deadlock, crash or a random error message depending where the collision occurred. In CI, you would usually notice this bug as a build job timing out or random tests failing for random reasons.

I have attached a repro below:

[_repro.zip](https://github.com/nightroman/Invoke-Build/files/6211497/_repro.zip)

Please read the `README.md` in the attachment for usage instructions. The task names should be self-explanatory.

Keep in mind, that the issue replication might require multiple runs in order to get an actual deadlock or exception. A successful run does not mean there were no multiple threads executing in the same runspace. It may simply mean they did not collide with each other. Also, the repro is rather simple and short-lived, so the possibility of collision is small. But in larger tasks, which is the use case for parallelization, the problem becomes more prevalent.

The issue has been replicated in `Windows PowerShell 5.1`, `PSCore 7.1 on Windows` and `PSCore 7.1 on Linux under WSL2` however the repro has only been tested on Windows.

A potential solution has also been provided in the `task Succeeds-ParallelStartJob` which instead of using `BeginInvoke` uses PowerShell's `Start-Job`. The implementation is rudimentary and should act only as a proof of concept.

Please, let me know what you think.

**Comment by nightroman** [comment]
@codepic Thank you for the detailed report. I'll look at this closely. Before doing this, I have some doubts about using `Start-Job`, the idea of `Build-Parallel` was to work in the same process without using `Start-Job`...

**Comment by codepic** [comment]
@nightroman I understand. Maybe this can be worked around somehow. I don't know 😕  I will let you know if I find something.

**Comment by codepic** [comment]
Probably not a good idea to completely replace the current implementation only to potentially open another can of worms which will be found later. Perhaps a switch f.ex. `Build-Parallel -Parallelize InProcess|ByThread|ByProcess` or similar could do defaulting to the current implementation `InProcess`.

Also, `Start-Job` isn't always available as per [the docs](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/start-job) and it suggests to use [Start-ThreadJob](https://docs.microsoft.com/en-us/powershell/module/threadjob/start-threadjob) in this scenario instead.

**Comment by nightroman** [comment]
FWIW. With my local IB (not your provided bootstrap) it keeps failing consistently with this error.

```
ERROR: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
At C:\Users\roman.kuzmin\Desktop\_210326_ka-ib_repro\.build.ps1:62 char:5
+     Write-Build Magenta ([Repro]::GetRepro() | Out-String)
```

**Comment by nightroman** [comment]
One more thing... You are using classes. If this problem turns out to be about classes then I do not want to look at it. Sorry. I find PowerShell classes not reliable. (Script modules, too, by the way). All in all, can we repro this problem without classes?

**Comment by nightroman** [comment]
Another thought (do not get me wrong, I am not trying to say there is no issue or it's not my issue). The mentioned issue talks about problems on using `InvokeWithPipe`. But `Build-Parallel.ps1` does not seem to use it (maybe implicitly?)

**Comment by codepic** [comment]
> All in all, can we repro this problem without classes?

Try by replacing the `Task task` in repro with this:

```
task Task {
    function fibonacci {
        param([int]$n)
        [bigint]$a=0
        [bigint]$b=1
        foreach ($x in 0..$n)
        {
            $a,$b = $b,($a+$b)
        }
        $b
    }
    fibonacci 100000 | % {'complete'}
}
```

At least for me it seems to deadlock pretty consistently at around task `#32`.

**Comment by codepic** [comment]
> The mentioned issue talks about problems on using `InvokeWithPipe`. But `Build-Parallel.ps1` does not use it. Or maybe it does, implicitly?

I had a look and it seems that `ScriptBlock` internally and eventually calls [InvokeWithPipe](https://github.com/PowerShell/PowerShell/blob/15e4e8a6af2f120bbd1eef8607131d878677c0ad/src/System.Management.Automation/engine/lang/scriptblock.cs#L940).

**Comment by nightroman** [comment]
[#4003]: https://github.com/PowerShell/PowerShell/issues/4003

Right... Where are we?

We know that `Build-Parallel` has problems in scenarios with PowerShell
classes. We have the reproducible case.

Solution? Acknowledge, document this, add the note: do not use `Build-Parallel`
with PowerShell classes involved. We do not know how to work around these
PowerShell engine and PowerShell classes problems.

Like I said, I do not want to engage `Start-Job`. The idea of `Build-Parallel`
was not using `Start-Job`, to run parallel builds in the same process, with
unique features like sharing live data and environment.

If one wants using `Start-Job` then `Build-Parallel` is not needed as such:

```
@(
    Start-Job {Invoke-Build Task1 $args[0]} -ArgumentList <script1>
    Start-Job {Invoke-Build Task2 $args[0]} -ArgumentList <script2>
    ... # add more builds
) | Wait-Job | Receive-Job
```

About `Start-ThreadJob`. It might be more attractive for its same process
nature. But it is not available in PowerShell 2-5.1 which IB supports and,
just like `Start-Job`, it may be just used directly without `Build-Parallel`.
In other words, I would not like to engage it in `Build-Parallel`.

Without classes. Do we have a reproducible case? The issue [#4003] mentions
some scenarios with a script block passed to new runspaces and invoked by
`InvokeReturnAsIs`. I tried and could not yet see a problem. Can you?
It may depend on machine, environment, etc. of course. See attached.

[3-script-arg.zip](https://github.com/nightroman/Invoke-Build/files/6215369/3-script-arg.zip)

**Comment by codepic** [comment]
Sounds good to me. I totally understand you don't want to code your solution according to some bugs in another one.

I will look if I can repro this without classes and you'll be the first to know. I will also look into the zip you provided.

**Comment by nightroman** [comment]
NB Added the note about avoiding classes:
- to wiki https://github.com/nightroman/Invoke-Build/wiki/Parallel-Builds
- to `Parallel-Build` help (to be in the next release)

**Comment by nightroman** [comment]
I think we may close at this point. Feel free to reopen with new input. Thank you for reporting and investigating once again.
